### PR TITLE
Factor out timeout and coverage handling to make the InterleavedSearc…

### PR DESCRIPTION
**Pull request by baldersheim**
…hInvoker easier to understand and modify.

@vekterli or @bjorncs or @bratseth PR

### CODE REVIEW

**Review by bratseth — APPROVED** [review]
Nice!

**Review by vekterli — COMMENTED** [review]
:+1: some minor questions/comments

File: container-search/src/main/java/com/yahoo/search/dispatch/InterleavedSearchInvoker.java [hunk 1]
```diff
@@ -40,38 +36,35 @@ public class InterleavedSearchInvoker extends SearchInvoker implements ResponseM
...
-    private long answeredActiveDocs = 0;
-    private long answeredTargetActiveDocs = 0;
-    private int askedNodes = 0;
-    private int answeredNodes = 0;
-    private int answeredNodesParticipated = 0;
-    private boolean timedOut = false;
-    private boolean degradedByMatchPhase = false;
-
-    public InterleavedSearchInvoker(Collection<SearchInvoker> invokers,
+    TimeoutHandler timeoutHandler;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/InterleavedSearchInvoker.java:48]: Yes

File: container-search/src/main/java/com/yahoo/search/dispatch/InterleavedSearchInvoker.java [hunk 2]
```diff
@@ -40,38 +36,35 @@ public class InterleavedSearchInvoker extends SearchInvoker implements ResponseM
...
         this.invokers = Collections.newSetFromMap(new IdentityHashMap<>());
         this.invokers.addAll(invokers);
         this.searchCluster = searchCluster;
         this.group = group;
         this.availableForProcessing = newQueue();
         this.alreadyFailedNodes = alreadyFailedNodes;
+        coverageAggregator = new CoverageAggregator(invokers.size());
+    }
+
+    TimeoutHandler create(DispatchConfig config, int askedNodes, Query query) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/InterleavedSearchInvoker.java:64]: Fixed

[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/InterleavedSearchInvoker.java:64]: Yes

File: container-search/src/main/java/com/yahoo/search/dispatch/CoverageAggregator.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        answeredTargetActiveDocs += source.getTargetActive();
+        answeredNodesParticipated += source.getNodes();
+        answeredNodes++;
+        degradedByMatchPhase |= source.isDegradedByMatchPhase();
+        timedOut |= source.isDegradedByTimeout();
+    }
+    public int getAskedNodes() {
+        return askedNodes;
+    }
+    public int getAnswerdNodes() {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/CoverageAggregator.java:48]: Fixed

File: container-search/src/main/java/com/yahoo/search/dispatch/CoverageAggregator.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        if (timedOut) {
+            degradedReason |= timeoutHandler.reason();
+        }
+        if (degradedByMatchPhase) {
+            degradedReason |= DEGRADED_BY_MATCH_PHASE;
+        }
+        coverage.setDegradedReason(degradedReason);
+        return coverage;
+    }
+    public CoverageAggregator adjustDegradedCoverage(int searchableCopies, TimeoutHandler timeoutHandler) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/CoverageAggregator.java:71]: Agree

File: container-search/src/main/java/com/yahoo/search/dispatch/TimeoutHandler.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.yahoo.search.dispatch;
+
+/**
+ * Computes next timeout
+ *
+ * @author baldersheim
+ */
+public interface TimeoutHandler {
+    long nextTimeout(int answeredNodes);
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/TimeoutHandler.java:9]: Both

File: container-search/src/main/java/com/yahoo/search/dispatch/TimeoutHandler.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+    int reason();
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/TimeoutHandler.java:10]: Fixed

**Pull request merged**